pkg/alert/rule_controller.go

```diff
@@ -0,0 +1,465 @@
+		return
+	}
+
+	klog.V(4).Infof("AlertingRule deleted: %s", key)
+	rc.enqueue(key)
+}
+
+// handleAlertingRuleUpdate handles update events for the AlertingRule informer.
+func (rc *RuleController) handleAlertingRuleUpdate(oldObj, newObj interface{}) {
+	key, ok := rc.keyFunc(newObj)
+	if !ok {
+		return
+	}
+
+	if !rc.inPlatformNamespace(newObj) {
+		klog.V(4).Infof("Ignoring update of AlertingRule %q in non-platform namespace", key)
+		return
+	}
+
+	// If the ResourceVersion hasn't changed, there's nothing to do.
+	if oldObj.(*osmv1alpha1.AlertingRule).ResourceVersion == newObj.(*osmv1alpha1.AlertingRule).ResourceVersion {
+		klog.V(4).Info("Skipping AlertingRule update due to identical ResourceVersion (%s)",
+			newObj.(*osmv1alpha1.AlertingRule).ResourceVersion)
+		return
+	}
+
+	// If the Generation hasn't changed, the spec hasn't changed. Nothing to do.
+	if oldObj.(*osmv1alpha1.AlertingRule).Generation == newObj.(*osmv1alpha1.AlertingRule).Generation {
+		klog.V(4).Infof("Skipping AlertingRule update due to identical Generation (%d)",
+			newObj.(*osmv1alpha1.AlertingRule).Generation)
+		return
+	}
+
+	klog.V(4).Infof("AlertingRule updated: %s", key)
+	rc.enqueue(key)
+}
+
+// worker starts processing of the controller's work queue.
+func (rc *RuleController) worker(ctx context.Context) {
+	for rc.processNextWorkItem(ctx) {
+	}
+}
+
+// processNextWorkItem processes the next item on the work queue.
+func (rc *RuleController) processNextWorkItem(ctx context.Context) bool {
+	key, quit := rc.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer rc.queue.Done(key)
+
+	if err := rc.sync(ctx, key.(string)); err != nil {
+		utilruntime.HandleError(errors.Wrap(err,
+			fmt.Sprintf("Error syncing AlertingRule (%s)", key.(string))))
+
+		// Re-queue failed sync.
+		rc.queue.AddRateLimited(key)
+
+		return true
+	}
+
+	klog.V(4).Infof("AlertingRule successfully synced: %s", key.(string))
+	rc.queue.Forget(key) // Reset rate-limiting.
+
+	return true
+}
+
+// sync reconciles the desired state of the AlertingRule for the given key.  It
+// fetches the AlertingRule from the API, then generates a new PrometheusRule
+// object based on the AlertingRule.
+func (rc *RuleController) sync(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	rule, err := rc.client.GetAlertingRule(ctx, namespace, name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// Deletion is handled automatically via the OwnerReference.
+		return nil
+
+	case err != nil:
+		return err
+	}
+
+	klog.V(4).Infof("Syncing AlertingRule: %s", key)
+
+	rule.APIVersion = osmv1alpha1.GroupVersion.String()
+	rule.Kind = "AlertingRule"
+
+	// Generate the new or updated PrometheusRule object.
+	promRule := &monv1.PrometheusRule{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rc.promRuleName(namespace, name),
+			Namespace: rc.client.Namespace(),
+			OwnerReferences: []metav1.OwnerReference{
```

Comment:
Now I remember that owner reference can't cross namespaces as explained in https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications
:thinking: not sure how to solve this

Reply:
Responded [here](https://github.com/openshift/cluster-monitoring-operator/pull/1590#issuecomment-1073784960).